crates/task-impls/src/consensus2/handlers.rs

```diff
@@ -133,6 +133,23 @@ pub(crate) async fn handle_view_change<TYPES: NodeType, I: NodeImplementation<TY
     // Move this node to the next view
     task_state.cur_view = new_view_number;
 
+    // If we have a decided upgrade certificate, we may need to upgrade the protocol version on a
+    // view change.
+    let decided_upgrade_certificate_read =
+        task_state.decided_upgrade_certificate.read().await.clone();
+    if let Some(cert) = decided_upgrade_certificate_read {
+        if new_view_number == cert.data.new_version_first_view {
+            warn!(
+                "Updating version based on a decided upgrade cert: {:?}",
+                cert
+            );
+            let mut version = task_state.version.write().await;
+            let new_version = cert.data.new_version;
+            *version = new_version;
+            broadcast_event(Arc::new(HotShotEvent::VersionUpgrade(new_version)), sender).await;
```

Comment:
I think most of this is (hopefully) not used anywhere, except the log which is helpful to see. Ideally we can remove it (possibly in a later PR).

Do you happen to know if the upgrade fails without updating `version`? I can investigate if not.

Reply:
Removing the version update doesn't break the upgrade test but the `example-upgrade` script does fail, so I'll keep it as is for now. We have this issue to improve the version trakcing: https://github.com/EspressoSystems/HotShot/issues/3378.